### ISSUE TITLE
Unable to build on Raspberry Pi 4 (aarch64)

### ISSUE THREAD

**Issue by vanfanel** [bbbradsmith/hatariB#11]
Hi there, Brad

I just discovered this new Hatari port and I have high hopes on it! Thanks for the work you're putting into this!

However, I am trying to build it on my Pi4 in aarch64 mode, and it fails to do so:

```
 CC     build/SDL.lo
gcc: fatal error: Killed signal terminated program cc1
compilation terminated.
gcc: fatal error: Killed signal terminated program cc1
compilation terminated.
gcc: fatal error: Killed signal terminated program cc1
compilation terminated.
make[2]: *** [Makefile:290: build/SDL_list.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:362: build/SDL_windowevents.lo] Error 1
make[2]: *** [Makefile:368: build/SDL_rwops.lo] Error 1
```

I am puzzled that this builds SDL2, since this is supposed to be a Libretro port.
Maybe I am building it wrong? I did:

`make full`

and since this Pi only has 2GB of RAM, I also tried:

`make -j1 full`

but I get the same errors. Any advice? Maybe it's early days, but I am so willing to try this and have some Captain Blood fun!

**Comment by bbbradsmith** [comment]
Hm, `gcc: fatal error: Killed signal terminated program cc1` doesn't seem to indicate any direct problem with the code, but rather that gcc itself can't manage for one reason or another. Possibly out of memory?

Since it's failed in the SDL build rather than hatariB's code, I would suggest trying to incrementally continue the SDL build by running its make directly. Not with `make full`, which is a clean and rebuild, but instead, after the point of failure, do `cd SDL/build` and then `make`. Then finally `make install` if it succeeds. If that completes, then you can just run `make` in the hatariB directory to build that without restarting the SDL build.

The problem might actually be that in `makefile.sdl` I added the `-j` flag directly. Possibly the environment just can't handle building SDL2 in parallel like this? Maybe I should do something to make this optional... though in most cases it's desirable because it will speed up the build considerably.

The "heaviest" part of the compile should actually come a big later when compiling the hatari procedurally generated UAE CPU files, which are pretty huge and take a lot of memory and CPU time to build.

If you're curious why SDL2 is a dependency, there is some information here: [DEVELOP.md # SDL2 Usage](https://github.com/bbbradsmith/hatariB/blob/main/DEVELOP.md#sdl2-usage)

**Comment by bbbradsmith** [comment]
Though, I would also ask, is the aarch64 nightly build usable? [link](https://nightly.link/bbbradsmith/hatariB/workflows/build/main)

I don't have an RPi4 to test it with, myself. The Raspberry Pi OS 32-bit build does work for me on an older RPi I have, and the 64-bit build is configured exactly the same except replacing the gnu eabihf compiler with the aarch64 one.

**Comment by bbbradsmith** [comment]
I've added a `MULTITHREAD` option to the makfiles. Would you be willing to try the full build again but setting that variable to empty on the command line?

`make full MULTITHREAD=`

**Comment by vanfanel** [comment]
@bbbradsmith Yes, nightly build is perfectly usable. CPU usage sits around ~35% (this is on KMS/DRM and Wayland, I don't use old and broken X11 on the Pi4).
![st1](https://github.com/bbbradsmith/hatariB/assets/837585/883552d0-7089-4ade-b65e-f655dfab8de1)
![st2](https://github.com/bbbradsmith/hatariB/assets/837585/71fb354b-2869-4a6e-8c45-a164e6fa2a19)

Now let's try that new build option. But you shouldn't specify any job number on Makefiles: it's up to the user to specify that via "make -jn"

**Comment by vanfanel** [comment]
@bbbradsmith Could you please add an option to build against system-installed SDL2?
That way, building would be way faster, and in most cases SDL2 is a very common lib to have installd system-wide in any GNU/Linux system.

**Comment by vanfanel** [comment]
Oh, it fails anyway:

```
/root/src/libretro/hatariB/SDL/src/joystick/SDL_joystick.c:1465:13: error: ‘UpdateEventsForDeviceRemoval’ defined but not used [-Werror=unused-function]
 1465 | static void UpdateEventsForDeviceRemoval(int device_index, Uint32 type)
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

**Comment by bbbradsmith** [comment]
Thank you for testing the nightly build. It helps to have verification.

Using a globally installed SDL2 was the initial approach, but it led to various problems. It did work on some platforms, but not on others, and also had a capacity to conflict with RetroArch's SDL2 drivers, if the user was using them. Building them directly was the only way to get a static version usable on all platforms. I've reduced its configuration to the absolute minimum, at least, but yes it becomes overhead for the full build.

If you wish, you can still provide your own by overriding the path variables: `SDL2_INCLUDE`, `SDL2_LIB` and `SDL2_LINK`.

The failure is strange, though, because I don't know how `-Werror=unused-function` gets added for your build. On all 6 nightly builds there is `-Wunused-function` indicating a warning there, but not an error.

Did SDL's config enable this on your platform, or is there something in your environment providing that?

**Comment by bbbradsmith** [comment]
You could try adding `--enable-werror=no` to `makefile.sdl` to see if that corrects it?

**Comment by vanfanel** [comment]
@bbbradsmith In the end this small SDL error was my fault. I was doing:
`CFLAGS="-march=native -mtune=native -fomit-frame-pointer"  CXXFLAGS="-march=native -mtune=native -fomit-frame-pointer" make full MULTITHREAD=`

After removing my custom CFLAGS and CXXFLAGS, it builds.

Next I will do a build using system-installed SDL2 as per your instructions.

**Comment by vanfanel** [comment]
@bbbradsmith How should I pass my own custom CFLAGS? I don't see any env var for extra flags on the Makefile.

**Comment by vanfanel** [comment]
What I am trying is to simply pass custom CFLAGS to the Hatari core, ideally both the core and SDL should receive the ones I pass.
But they seem to break compilation somehow.

**Comment by vanfanel** [comment]
@bbbradsmith I am doing this to try to build against system SDL2:
`
make full MULTITHREAD= SDL2_INCLUDE=/usr/local/include/SDL2 SDL2_LIB=/usr/local/lib SDL2_LINK=/usr/local/lib`

..but the build system is still trying to build SDL2. Any advices, please?

**Comment by vanfanel** [comment]
@bbbradsmith Ah no, passing `SDL2_LINK=/usr/local/bin/sdl2-config` results in file format not recognized.

This seems to work perfectly instead:
`make MULTITHREAD= SDL2_INCLUDE=/usr/local/include/SDL2 SDL2_LINK=/usr/local/lib/libSDL2.a`

**Comment by bbbradsmith** [comment]
Hmm, I think the answer you be to run `/usr/local/bin/sdl2-config` and copy its result into that environment variable. It's a command line script that generates some linker arguments text. (The makefile's default actually runs it and uses the result.)

**Comment by bbbradsmith** [comment]
But yeah, just linking the library should probably be sufficient. I'm still strongly considering just replacing SDL2 altogether, it just needs some careful consideration.

Thanks so much for testing things!

**Comment by vanfanel** [comment]
@bbbradsmith My pleasure! I never though I was going to help test another work of the great Lizard NES game! 

Replacing SDL2 is a good idea, indeed.

**Comment by bbbradsmith** [comment]
Also I should really play Rodland and Captain Blood. ;)

**Comment by vanfanel** [comment]
@bbbradsmith Of course! Don't forget to have fun with your own creation! Amberstar is another GREAT game on the ST, if you're into european RPGs (I love them!).

**Comment by vanfanel** [comment]
In the end, that doesn't even work:

```
cc -o build/hatarib.so -shared -Wall -Werror -static-libgcc build/core/core.o build/core/core_file.o build/core/core_input.o build/core/core_disk.o build/core/core_config.o build/core/core_osk.o hatari/build/src/libcore.a hatari/build/src/falcon/libFalcon.a hatari/build/src/cpu/libUaeCpu.a hatari/build/src/gui-sdl/libGuiSdl.a hatari/build/src/libFloppy.a hatari/build/src/debug/libDebug.a hatari/build/src/libcore.a /usr/lib/x86_64-linux-gnu/libz.a /usr/local/lib/libSDL2.a
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libz.a(deflate.o): warning: relocation against `z_errmsg' in read-only section `.text'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libz.a(zutil.o): relocation R_X86_64_PC32 against symbol `z_errmsg' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make: *** [makefile:118: build/hatarib.so] Error 1

```

Really, system-wide zlib should be used when no "full" is passed to make.

**Comment by bbbradsmith** [comment]
That last error is that your /usr/lib/x86_64-linux-gnu/libz.a is not built as relocatable code (-fPIC), so it can't be linked into a shared object.

That seems unusual to me.The libz.a I'm used to finding in package distributions is a relocatable stub that implicitly imports the system-installed zlib shared object. Yours looks like one instead built for static linking into executables? (i.e. `configure --static`) I'm a little curious how that version was installed, and/or whether it's different than what you'd get with `-lz`.

I'm more curious about what you meant 2 comments back, though. You said you had problems building because it tries to link against the local zlib? What was the problem?

**Comment by bbbradsmith** [comment]
(Sorry, didn't mean to close, that was an accidental click. I will close it when the conversation seems finished, though I'll probably link to it in the readme about build instructions.)

**Comment by vanfanel** [comment]
@bbbradsmith Honestly I have been doing so many build tries that I don't know what I meant there, sorry.

**Comment by vanfanel** [comment]
@bbbradsmith Ah, now I remember. When trying to build against zlib .so, like this:

make MULTITHREAD= SDL2_INCLUDE=/usr/local/include/SDL2 SDL2_LINK=/usr/local/lib/libSDL2.a ZLIB_INCLUDE=/usr/include ZLIB_LINK=/usr/lib/x86_64-linux-gnu/libz.so

I get:
```

/usr/bin/ld: /usr/local/lib/libSDL2.a(SDL_assert.o): warning: relocation against `stdin@@GLIBC_2.2.5' in read-only section `.text'
/usr/bin/ld: /usr/local/lib/libSDL2.a(SDL_dynapi.o): relocation R_X86_64_PC32 against symbol `stderr@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make: *** [makefile:118: build/hatarib.so] Error 1
```

So, for now, I am unable to build on x86_64, unless you give me more clues.

**Comment by bbbradsmith** [comment]
Well, I think that's the same problem as your zlib. You have a static libSDL2 but it wasn't built with -fPIC, so it can't be linked into a shared object. It's made for stand-alone exectuables only.

So, you'd need to link with a shared object version of libSDL2 rather than this static one, or you'd need to build your own static libSDL2 with -fPIC. I think in this case you would want the shared library.


This is actually one of the problems I was addressing when I decided to vendor them and build them from scratch. The shared object versions just _aren't_ viable for some target platforms, and the static libraries, which aren't universally available either, aren't made to link with shared objects. There were other issues too, but this was one of them.

**Comment by vanfanel** [comment]
@bbbradsmith Totally right on that one! You are right: static libs seem to be built without -fPIC here, so it's a matter of using the shared object versions. Sorry I seemed so obtuse yesterday, I was lacking some sleep hours :)

So... this works:
`make MULTITHREAD= SDL2_INCLUDE=/usr/local/include/SDL2 SDL2_LINK=/usr/local/lib/libSDL2.so ZLIB_INCLUDE=/usr/include ZLIB_LINK=/usr/lib/x86_64-linux-gnu/libz.so`

**Comment by vanfanel** [comment]
Here's my old laptop pretending to be an ST thanks to this fantastic core, shader and all! (Shadowmask, scanlines, curvature... sweet!)

![atarist-](https://github.com/bbbradsmith/hatariB/assets/837585/3a70a823-e108-4598-9497-f4bd005480b2)

**Comment by bbbradsmith** [comment]
Ah, I didn't consider the questions obtuse at all. Setting up builds is a nightmare and it took me years to even begin to find my way around them... and when multiple platforms are involved the rules change for each one. :S

Anyway, it seems okay to close this one now, but feel free to re-open it or make a new issue if you think there is further stuff. I will add some notes to the readme pursuant to the things we discussed.

Glad to see that little green desktop in action. ;)
